Improve sidebarAction API test coverage.

AriYoung00 opened the pull request:
#### 0f149866988c4425bb777f4747bf9273a5a2aa5e
<pre>
Improve sidebarAction API test coverage.
<a href="https://webkit.org/b/278395">https://webkit.org/b/278395</a>
<a href="https://rdar.apple.com/134355062">rdar://134355062</a>

Reviewed by Timothy Hatcher.

This PR improves test coverage for the sidebarAction JavaScript
extensions API, especially as pertaining to the semantics of the API
methods relating to the panel URL and panel title. It also adds several
drive-by fixes for bugs I found while writing these tests.

* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPISidebarCocoa.mm:
(WebKit::WebExtensionContext::sidebarSetTitle): Actually set title on
sidebar object in this method
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::getSidebar): Return nullopt rather than
default sidebar when no sidebar is found.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionSidebarCocoa.mm:
(WebKit::getDefaultSidebarTitleFromExtension): Default to extension
display name when no title is found in manifest
(WebKit::WebExtensionSidebar::extensionContext const): Switch to WeakPtr
for holding a reference to the current extension context
(WebKit::WebExtensionSidebar::parent const): Add nil check before we
attempt to get the sidebar for *(tab-&gt;window()), use extensionContext()
for context validity check rather than m_context
(WebKit::WebExtensionSidebar::setTitle): When setting the title to
std::nullopt on the default sidebar, retrieve manifest-default display
name instead.
(WebKit::WebExtensionSidebar::setSidebarPath): When setting the path to
std::nullopt on the default sidebar, retrieve manifest-default sidebar
path instead.
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPISidebarActionCocoa.mm:
(WebKit::parseDetailsStringFromKey): Return String variant in happy case
(was mistakenly returning SidebarError instead).
(WebKit::WebExtensionAPISidebarAction::setPanel): Set panelPath argument
in message to panel path rather than std::nullopt.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPISidebar.mm:
(TestWebKitAPI::WKWebExtensionAPISidebar::getManagerFor): Add helper
method to set up TestWebExtensionManager.
(TestWebKitAPI::TEST_F): Add assorted tests exercising the title and
panel URL semantics of the sidebarAction API, including setting,
modifying, retrieving, and clearing these &quot;properties.&quot;
* Source/WebKit/UIProcess/Extensions/WebExtensionSidebar.h: Switch to
WeakPtr for holding a reference to the current WebExtensionContext,
rather than WeakRef, rename m_context to m_extensionContext

Canonical link: <a href="https://commits.webkit.org/282533@main">https://commits.webkit.org/282533@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fc3abc6b607bfa6b42f3e02f8b657677473e12c5

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/63421 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/42777 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/16018 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/67442 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/59/builds/14029 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/65541 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/50465 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/14309 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/67442 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wincairo-tests~~](https://ews-build.webkit.org/#/builders/59/builds/14029 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/66490 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/39713 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/54923 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/67442 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/36390 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/12280 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/12901 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/57922 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/63/builds/12604 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/69138 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/87/builds/7368 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/12191 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/69138 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/86/builds/7400 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/54998 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/69138 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-1-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/9584 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/38598 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/39677 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/40789 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/39420 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionSidebarCocoa.mm:
```diff
@@ -101,7 +102,10 @@
+        .and_then([this](auto const& tab) -> std::optional<Ref<WebExtensionSidebar>> {
+            return tab->window() ? m_context->getSidebar(*(tab->window())) : std::nullopt;
+        })
         .value_or(m_context->defaultSidebar());
```
```suggestion
    return m_tab.and_then([this](auto const& tab) -> std::optional<Ref<WebExtensionSidebar>> {
            return tab->window() ? m_context->getSidebar(*(tab->window())) : std::nullopt;
        }).value_or(m_context->defaultSidebar());
```

xeenon commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPISidebar.mm:
```diff
@@ -113,6 +115,7 @@
     },
 };
 
+#pragma mark - Test Fixture
```
```suggestion
#pragma mark - Test Fixture

```

xeenon commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPISidebar.mm:
```diff
@@ -219,6 +225,514 @@
+
+        @"browser.test.assertDeepEq(await browser.sidebarAction.getTitle({ windowId: window1.id }), 'window-specific title')",
+        @"browser.test.assertDeepEq(await browser.sidebarAction.getTitle({}), 'global title')",
+        @"browser.test.assertDeepEq(await browser.sidebarAction.getTitle({ windowId: window2.id }), 'global title')",
```
`browser.test.assertEq` is better for strings and non-object types, unless you had another reason for `assertDeepEq`? `assertDeepEq` is kind of a hack currently…
```suggestion
        @"browser.test.assertEq(await browser.sidebarAction.getTitle({ windowId: window1.id }), 'window-specific title')",
        @"browser.test.assertEq(await browser.sidebarAction.getTitle({}), 'global title')",
        @"browser.test.assertEq(await browser.sidebarAction.getTitle({ windowId: window2.id }), 'global title')",
```

AriYoung00 replied:
I'll go ahead and switch all of those over -- I only used `assertDeepEq` to stay consistent with my habit of always using `===` instead of `==` when writing JS.

xeenon commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPISidebar.mm:
Awesome tests!

AriYoung00 replied:
Thank you!

kiaraarose commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -2659,18 +2659,18 @@ static WKWebExtensionContextError toAPI(WebExtensionContext::Error error)
 {
-    if (auto *windowAction = m_sidebarWindowMap.get(window))
-        return *windowAction;
+    if (auto *windowSidebar = m_sidebarWindowMap.get(window))
```
i think it's preferred to use `RefPtr` instead of `auto *` here. we had feedback previously to not use `auto` for `Ref` and `RefPtr` types

AriYoung00 replied:
In this case, although `m_sidebarWindowMap` has value type `Ref<WebExtensionSidebar>`, it seems as though `m_sidebarWindowMap.get` is returning a raw `WebExtensionSidebar *`. Would it still be advantageous to specify the full type here?

kiaraarose replied:
hmm i'm not sure why that is. cc @xeenon

xeenon replied:
It would be best to do:
```c++
if (RefPtr windowSidebar = m_sidebarWindowMap.get(window))
```
So the local holds a reference and it can't disappear from under you. This is an area we don't do well with everywhere yet.

AriYoung00 replied:
Odd, I swear that was generating a compiler error for me about 30 minutes ago, but it seems to have resolved itself / I was doing something wrong before. Adding this change now.

kiaraarose commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -2659,18 +2659,18 @@ static WKWebExtensionContextError toAPI(WebExtensionContext::Error error)
 {
-    if (auto *tabAction = m_sidebarTabMap.get(tab))
-        return *tabAction;
+    if (auto *tabSidebar = m_sidebarTabMap.get(tab))
```
ditto

kiaraarose commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionSidebarCocoa.mm:
```diff
@@ -101,8 +102,9 @@
-    return m_tab.and_then([this](auto const& tab) { return m_context->getSidebar(*tab->window()); })
-        .value_or(m_context->defaultSidebar());
+    return m_tab.and_then([this](auto const& tab) -> std::optional<Ref<WebExtensionSidebar>> {
+            return tab->window() ? m_context->getSidebar(*(tab->window())) : std::nullopt;
```
i think `m_extensionContext` would be a more precise name for this variable

kiaraarose replied:
or rather your `extensionContext()` method

kiaraarose replied:
and also, why does that method return an optional context?

AriYoung00 replied:
It returns an `optional<Ref<WebExtensionContext>>` since we hold a weak pointer to the current context as a class member, so we need to check if it's still there on every access -- I just use the optional to indicate whether it's still there. In this particular instance, I'm using the member variable to access context rather than the getter because I check to ensure that context is still there at the beginning of this method, so there's no need to re-run the check by accessing context through the getter later on.

AriYoung00 committed:
Improve sidebarAction API test coverage.
https://webkit.org/b/278395
rdar://134355062

Reviewed by Timothy Hatcher.

This PR improves test coverage for the sidebarAction JavaScript
extensions API, especially as pertaining to the semantics of the API
methods relating to the panel URL and panel title. It also adds several
drive-by fixes for bugs I found while writing these tests.

* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPISidebarCocoa.mm:
(WebKit::WebExtensionContext::sidebarSetTitle): Actually set title on
sidebar object in this method
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::getSidebar): Return nullopt rather than
default sidebar when no sidebar is found.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionSidebarCocoa.mm:
(WebKit::getDefaultSidebarTitleFromExtension): Default to extension
display name when no title is found in manifest
(WebKit::WebExtensionSidebar::extensionContext const): Switch to WeakPtr
for holding a reference to the current extension context
(WebKit::WebExtensionSidebar::parent const): Add nil check before we
attempt to get the sidebar for *(tab->window()), use extensionContext()
for context validity check rather than m_context
(WebKit::WebExtensionSidebar::setTitle): When setting the title to
std::nullopt on the default sidebar, retrieve manifest-default display
name instead.
(WebKit::WebExtensionSidebar::setSidebarPath): When setting the path to
std::nullopt on the default sidebar, retrieve manifest-default sidebar
path instead.
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPISidebarActionCocoa.mm:
(WebKit::parseDetailsStringFromKey): Return String variant in happy case
(was mistakenly returning SidebarError instead).
(WebKit::WebExtensionAPISidebarAction::setPanel): Set panelPath argument
in message to panel path rather than std::nullopt.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPISidebar.mm:
(TestWebKitAPI::WKWebExtensionAPISidebar::getManagerFor): Add helper
method to set up TestWebExtensionManager.
(TestWebKitAPI::TEST_F): Add assorted tests exercising the title and
panel URL semantics of the sidebarAction API, including setting,
modifying, retrieving, and clearing these "properties."
* Source/WebKit/UIProcess/Extensions/WebExtensionSidebar.h: Switch to
WeakPtr for holding a reference to the current WebExtensionContext,
rather than WeakRef, rename m_context to m_extensionContext

Canonical link: https://commits.webkit.org/282533@main
